### PR TITLE
Avoid leak of internal implementation in tasty.Reflection

### PR DESCRIPTION
**Pull request by nicolasstucki**
* Remove `Reflection.internal`: the source of the leak
* Add `CompilerInterface` as a self type
* Add common type abstraction for `Reflection` and `ComplerInterface`
* Add `ComplerInterface.leak` to allow access to internals within the library

### CODE REVIEW

File: library/src-bootstrapped/scala/quoted/Expr.scala [hunk 1]
```diff
@@ -75,7 +75,8 @@ object Expr {
    *   Otherwise returns `expr`.
    */
   def betaReduce[T](expr: Expr[T])(using qctx: QuoteContext): Expr[T] =
-    qctx.tasty.internal.betaReduce(expr.unseal) match
+    val qctx2 = scala.internal.tasty.CompilerInterface.leaked(qctx)
```
[the remark this answers is not in the document]
Comment (liufengyun) [reply library/src-bootstrapped/scala/quoted/Expr.scala]: What about just return an instance of `CompilerInterface`? That will make things more clean:

```Scala
val ci: CompilerInterface = scala.internal.tasty.CompilerInterface(qtx)
```

**Review by liufengyun — APPROVED** [review]
Otherwise, LGTM

File: library/src-bootstrapped/scala/quoted/Expr.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply library/src-bootstrapped/scala/quoted/Expr.scala]: I tried that, but we lose the relation between `qctx.tasty.Tree` and `ci.Tree`. It is also something that we will not have to add often (or ever) in the future. There are some uses that may be removed if we add the functionality to the `Reflection` interface. As that would change the user interface I would prefer to do it in its own self contained PR.

[the remark this answers is not in the document]
Comment (liufengyun) [reply library/src-bootstrapped/scala/quoted/Expr.scala]: What about the following:

```Scala
object CompilerInterface {
  def apply(qctx: QuoteContext): scala.internal.tasty.CompilerInterface & qctx.tasty.type =  qctx.tasty
}
```

[the remark this answers is not in the document]
Comment (nicolasstucki) [reply library/src-bootstrapped/scala/quoted/Expr.scala]: That was what I tried.

```scala
[error] -- [E007] Type Mismatch Error: /Users/nicolasstucki/GitHub/dotty2/library/src-bootstrapped/scala/quoted/Expr.scala:80:18 
[error]         val ci = CompilerInterface(qctx)
[error] 80 |    ci.betaReduce(expr.unseal) match
[error]    |                  ^^^^^^^^^^^
[error]    |                  Found:    qctx.tasty.Term
[error]    |                  Required: ci.Term
```

[the remark this answers is not in the document]
Comment (liufengyun) [reply library/src-bootstrapped/scala/quoted/Expr.scala]: Might be a bug in TypeComparer? I'll try to minimize.

[the remark this answers is not in the document]
Comment (LPTK) [reply library/src-bootstrapped/scala/quoted/Expr.scala]: @liufengyun Dotty currently doesn't support reasoning about refining singletons, which feels like a big oversight: https://github.com/lampepfl/dotty-feature-requests/issues/89

[the remark this answers is not in the document]
Comment (liufengyun) [reply library/src-bootstrapped/scala/quoted/Expr.scala]: Thanks for the pointer @LPTK . Yes, indeed it is. We also lose the reasoning about paths for class parameters. I agree that a systematic approach is needed to support reasoning about path equality in path-dependent types.

**Pull request merged**